automation/workflow/workflow.go

```diff
@@ -0,0 +1,122 @@
+package workflow
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/google/uuid"
+	"github.com/qri-io/qri/profile"
+)
+
+var (
+	// ErrNotFound indicates that the workflow was not found in the store
+	ErrNotFound = fmt.Errorf("workflow not found")
+)
+
+// ID is a string identifier for a workflow
+type ID string
+
+// NewID creates a new workflow identifier
+func NewID() ID {
+	return ID(uuid.New().String())
+}
+
+// String returns the underlying id string
+func (id ID) String() string { return string(id) }
+
+// SetIDRand sets the random reader that NewID uses as a source of random bytes
+// passing in nil will default to crypto.Rand. This can be used to make ID
+// generation deterministic for tests. eg:
+//    myString := "SomeRandomStringThatIsLong-SoYouCanCallItAsMuchAsNeeded..."
+//    run.SetIDRand(strings.NewReader(myString))
+//    a := NewID()
+//    run.SetIDRand(strings.NewReader(myString))
+//    b := NewID()
+func SetIDRand(r io.Reader) {
+	uuid.SetRand(r)
+}
+
+// State holds all the information necessary to run a workflow
+type State struct {
+	ID        ID
+	DatasetID string
+	OwnerID   profile.ID
+	Deployed  bool
+}
+
+// Store manages & stores workflows, allowing listing and updating of workflows
+type Store interface {
+	Updater
+	// Lister
+}
+
+// An Updater can store and update workflows
+type Updater interface {
+	Create(did string, pid profile.ID) (*State, error)
+	Get(wid ID) (*State, error)
+	// Remove
+	// Update
+	// Deploy
+	// Undeploy
+}
+
+// A Lister lists entries from a workflow store
+type Lister interface {
+	// List
+	// ListDeployed
+}
+
+// NewDefaultStore returns the default expected store
+// TODO (ramfox): replace this with a file store asap, currently a MemStore
+func NewDefaultStore(ctx context.Context) (*MemStore, error) {
+	return NewMemStore(), nil
+}
+
+// MemStore is an in memory representation of a Store
+type MemStore struct {
+	mu        *sync.Mutex
+	workflows map[ID]*State
+}
+
+// NewMemStore return a MemStore
+func NewMemStore() *MemStore {
+	return &MemStore{
+		mu:        &sync.Mutex{},
+		workflows: map[ID]*State{},
+	}
+}
+
+// Create creates a new State and adds it to the Store. It does not check that
+// the given dataset or peer ids are valid, beyond that they are not empty
+func (m *MemStore) Create(did string, pid profile.ID) (*State, error) {
+	if did == "" {
+		return nil, fmt.Errorf("dataset ID required")
+	}
+	if pid == "" {
+		return nil, fmt.Errorf("profile ID required")
+	}
+	wf := &State{
+		ID:        NewID(),
+		DatasetID: did,
+		OwnerID:   pid,
+	}
+	m.mu.Lock()
+	m.workflows[wf.ID] = wf
+	m.mu.Unlock()
+	return wf, nil
+}
+
+// Get fetches a State using the associated ID
+func (m *MemStore) Get(wid ID) (*State, error) {
+	m.mu.Lock()
+	wf, ok := m.workflows[wid]
+	m.mu.Unlock()
+	if !ok {
+		return nil, ErrNotFound
+	}
+	return wf, nil
+}
+
+var _ Store = (*MemStore)(nil)
```

Comment:
move this to immodestly follow the `type MemStore struct` definition. Order I've come to prefer:

1. type definition
2. interface assertion(s)
3. constructors
4. Exported Methods
5. unexported methods